machine.go

```diff
@@ -1067,3 +1071,111 @@ func (m *Machine) CreateSnapshot(ctx context.Context, memFilePath, snapshotPath
 	m.logger.Debug("snapshot created successfully")
 	return nil
 }
+
+// setupBalloonDevice set up a balloon device if one does not exist
+func (m *Machine) CreateBalloonDevice(ctx context.Context, dev models.Balloon) error {
+	fmt.Println("Start to create a balloon device")
+	balloonConfig := models.Balloon{
+		AmountMb: Int64(0),
+		DeflateOnOom: Bool(false),
+		StatsPollingIntervals: 0,
+	}
+
+	if dev.AmountMb != nil {
+		balloonConfig.AmountMb = dev.AmountMb
+	}
+
+	if dev.DeflateOnOom != nil {
+		balloonConfig.DeflateOnOom = dev.DeflateOnOom
+	}
+
+	if dev.StatsPollingIntervals != 0 {
+		balloonConfig.StatsPollingIntervals = dev.StatsPollingIntervals
+	}
+
+	amountMb := *(balloonConfig.AmountMb)
+	m.logger.Infof("Install a balloon device with %v MB", amountMb)	
+	respNoContent, err := m.client.PutBalloon(ctx, &balloonConfig)
+
+	if err == nil {
+		m.logger.Printf("Installed balloon device with %v MB: %s", amountMb, respNoContent.Error())
+	} else {
+		m.logger.Errorf("Install balloon device failed : %s", err)
+	}
+	return err
+}
+
+// GetBalloonConfig gets the current balloon device configuration.
+func (m *Machine) GetBalloonConfig(ctx context.Context, v interface{}) error {
+	resp, err := m.client.DescribeBalloonConfig(ctx)
+	if err != nil {
+		m.logger.Errorf("Getting balloonConfig: %s", err)
+		return err
+	}
+
+	payloadData, err := json.Marshal(resp.Payload)
+	if err != nil {
+		m.logger.Errorf("Getting balloonConfig failed parsing payload: %s", err)
+		return err
+	}
+
+	if err := json.Unmarshal(payloadData, v); err != nil {
+		m.logger.Errorf("Getting balloonConfig failed parsing payload: %s", err)
+		return err
+	}
+
+	m.logger.Printf("GetBalloonConfig successful")
+	return nil
+}
+
+// UpdateBalloon will modify the current balloon with the new 
+// parameters of the ballonUpdate
+func (m *Machine) UpdateBalloon(ctx context.Context, amountMb int64, opts ...PatchBalloonOpt) error {
+	if amountMb < 0 {
+		m.logger.Printf("The amount of memory must be positive number")
+		return nil
+	}
+	if _, err := m.client.PatchBalloon(ctx, amountMb, opts...); err != nil {
+		m.logger.Errorf("UpdateBalloon failed: %v", err)
```

Comment:
here we want to check whether or not the error is related to config not being present and return the proper logging